File: app-java/src/main/java/com/TritiumGaming/phasmophobiaevidencepicker/views/account/AccountObtainCreditsView.java
package com.TritiumGaming.phasmophobiaevidencepicker.views.account;

import android.content.Context;
import android.content.res.ColorStateList;
import android.text.Html;
import android.util.AttributeSet;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatButton;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.constraintlayout.widget.ConstraintLayout;

import com.TritiumGaming.phasmophobiaevidencepicker.R;
import com.TritiumGaming.phasmophobiaevidencepicker.data.utilities.ColorUtils;
import com.TritiumGaming.phasmophobiaevidencepicker.firebase.firestore.transactions.user.FirestoreUser;
import com.google.firebase.auth.FirebaseUser;

public class AccountObtainCreditsView extends ConstraintLayout {

    private ColorStateList colorStateList;

    public AccountObtainCreditsView(@NonNull Context context) {
        super(context);

        init(getContext());
    }

    public AccountObtainCreditsView(@NonNull Context context, @Nullable @org.jetbrains.annotations.Nullable AttributeSet attrs) {
        super(context, attrs);

        init(getContext());
    }

    public AccountObtainCreditsView(@NonNull Context context, @Nullable @org.jetbrains.annotations.Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        init(getContext());
    }

    public AccountObtainCreditsView(@NonNull Context context, @Nullable @org.jetbrains.annotations.Nullable AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);

        init(getContext());
    }

    public void init(Context context) {
        inflate(context, R.layout.layout_account_obtain_credits, this);

        buildButtonStateColors(findViewById(R.id.button_ad_watch));
        buildButtonStateColors(findViewById(R.id.settings_account_buy_button));
        setWatchAdsLabelDescription();

        FirebaseUser user = null;
        try {
            user = FirestoreUser.getCurrentFirebaseUser();
        } catch (Exception e) {
            e.printStackTrace();
        }

        setContainerVisibility(user);
    }

    private void setContainerVisibility(@Nullable FirebaseUser user) {
        if(user == null) {
            setVisibility(GONE);
        } else {
            setVisibility(VISIBLE);
        }
    }

    public void setWatchAdsLabelDescription() {
        setWatchAdsLabelDescription(-1);
    }

    public void setWatchAdsLabelDescription(int quantity) {
        AppCompatTextView label_description = findViewById(R.id.label_ads_description);
        if(label_description == null) { return; }

        @ColorInt int color =
                ColorUtils.getColorFromAttribute(getContext(), R.attr.textColorPrimary);

        String descriptionQuantity = getResources().getQuantityString(
                R.plurals.marketplace_description_watch_ad,
                quantity, quantity);
        String colorHex = ColorUtils.intToHex(color);

        StringBuilder quantityStrTemp = new StringBuilder("<font color=")
                .append(colorHex).append(">");
        //String quantityStr = "";
        if(quantity >= 0) {
            quantityStrTemp.append(quantity);
        }
        quantityStrTemp.append("</font>");

        String descriptionFormat = getContext().getString(
                R.string.marketplace_description_watch_ad,
                quantityStrTemp.toString(), descriptionQuantity);
        label_description.setText(Html.fromHtml(descriptionFormat));
    }

    public void buildButtonStateColors(@Nullable AppCompatButton button) {
        if (button == null) {
            return;
        }

        ColorStateList colorStateList = button.getTextColors();
        @ColorInt int defaultColor = colorStateList.getDefaultColor();
        @ColorInt int disabledColor = ColorUtils.setColor(defaultColor, 50, -1, -1, -1);

        int[][] states = new int[][]{
                new int[]{android.R.attr.state_enabled},
                new int[]{-android.R.attr.state_enabled}
        };
        int[] colors = {
                defaultColor,
                disabledColor
        };
        colorStateList = new ColorStateList(states, colors);

        button.setTextColor(colorStateList);
    }

    public void enableAdWatchButton(boolean enable) {
        AppCompatButton button = findViewById(R.id.button_ad_watch);
        if(button == null) { return; }

        button.setEnabled(enable);
    }

}

File: app-legacy/src/main/java/com/tritiumgaming/phasmophobiaevidencepicker/presentation/ui/views/account/AccountObtainCreditsView.kt
package com.tritiumgaming.phasmophobiaevidencepicker.presentation.ui.views.account

import android.content.Context
import android.content.res.ColorStateList
import android.text.Html
import android.util.AttributeSet
import androidx.annotation.ColorInt
import androidx.appcompat.widget.AppCompatButton
import androidx.appcompat.widget.AppCompatTextView
import androidx.constraintlayout.widget.ConstraintLayout
import com.google.firebase.auth.FirebaseUser
import com.tritiumgaming.phasmophobiaevidencepicker.R
import com.tritiumgaming.phasmophobiaevidencepicker.data.remote.api.firestore.transactions.user.FirestoreUser.Companion.currentFirebaseUser
import com.tritiumgaming.phasmophobiaevidencepicker.util.ColorUtils.getColorFromAttribute
import com.tritiumgaming.phasmophobiaevidencepicker.util.ColorUtils.intToHex
import com.tritiumgaming.phasmophobiaevidencepicker.util.ColorUtils.setColor

class AccountObtainCreditsView : ConstraintLayout {

    constructor(context: Context) :
            super(context) { init(getContext()) }

    constructor(context: Context, attrs: AttributeSet?) :
            super(context, attrs) { init(getContext()) }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) :
            super(context, attrs, defStyleAttr) { init(getContext()) }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int) :
            super(context, attrs, defStyleAttr, defStyleRes) { init(getContext()) }

    fun init(context: Context?) {
        inflate(context, R.layout.layout_account_obtain_credits, this)

        findViewById<AppCompatButton>(R.id.button_ad_watch)?.let { b ->
            buildButtonStateColors(b)
        }
        findViewById<AppCompatButton>(R.id.settings_account_buy_button)?.let { b ->
            buildButtonStateColors(b)
        }
        setWatchAdsLabelDescription()

        var user: FirebaseUser? = null
        try { user = currentFirebaseUser }
        catch (e: Exception) { e.printStackTrace() }

        setContainerVisibility(user)
    }

    private fun setContainerVisibility(user: FirebaseUser?) {
        visibility = if (user == null) { GONE } else { VISIBLE }
    }

    private fun setWatchAdsLabelDescription() {
        setWatchAdsLabelDescription(-1)
    }

    fun setWatchAdsLabelDescription(quantity: Int) {
        val descriptionLabel =
            findViewById<AppCompatTextView>(R.id.label_ads_description) ?: return

        @ColorInt val color =
            getColorFromAttribute(context, R.attr.textColorPrimary)

        val descriptionQuantity = resources.getQuantityString(
            R.plurals.marketplace_description_watch_ad,
            quantity, quantity
        )
        val colorHex = intToHex(color)

        val quantityStrTemp = StringBuilder("<font color=")
            .append(colorHex).append(">")
        //String quantityStr = "";
        if (quantity >= 0) {
            quantityStrTemp.append(quantity)
        }
        quantityStrTemp.append("</font>")

        val descriptionFormat = context.getString(
            R.string.marketplace_description_watch_ad,
            quantityStrTemp.toString(), descriptionQuantity
        )
        descriptionLabel.text = Html.fromHtml(descriptionFormat)
    }

    private fun buildButtonStateColors(button: AppCompatButton) {

        var colorStateList = button.textColors
        @ColorInt val defaultColor = colorStateList.defaultColor
        @ColorInt val disabledColor = setColor(defaultColor, 50, -1, -1, -1)

        val states = arrayOf(
            intArrayOf(android.R.attr.state_enabled),
            intArrayOf(-android.R.attr.state_enabled)
        )
        val colors = intArrayOf(defaultColor, disabledColor)
        colorStateList = ColorStateList(states, colors)

        button.setTextColor(colorStateList)
    }

    fun enableAdWatchButton(enable: Boolean) {
        val button = findViewById<AppCompatButton>(R.id.button_ad_watch) ?: return

        button.isEnabled = enable
    }
}
